src/libraries/System.Private.CoreLib/src/System/Array.cs

```diff
@@ -808,10 +806,9 @@ public static void Fill<T>(T[] array, T value, int startIndex, int count)
                 ThrowHelper.ThrowCountArgumentOutOfRange_ArgumentOutOfRange_Count();
             }
 
-            for (int i = startIndex; i < startIndex + count; i++)
-            {
-                array[i] = value;
-            }
+            ref T first = ref Unsafe.Add(ref MemoryMarshal.GetArrayDataReference(array), startIndex);
```

Comment:
```suggestion
            ref T first = ref Unsafe.Add(ref MemoryMarshal.GetArrayDataReference(array), (nint)(uint)startIndex);
```
These casts [avoid](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABAJgEYBYAKGIAYACY8gOgCUBXAOwwEt8YLAMIR8AB14AbGFADKMgG68wMXAG4a9Jq049+ggJI8ZEMfKhKV6mpoDMTUgyEMA3jQYem92ADMGwAE8MGAYAWXIACkDggG0AXQZsKChsALQGXh4GXAwkjCMAExgEAEoGd083ak8ahl9/IJDomDYYPwBeOrawmHxoANCk3AALbEkWAHEYDABBZNSAEWxc1p8ZGC4VCKSUgJKNatqPYgB2Lr8AVS5cbDWWGYKCiPrm1fScvMLi/YqPAF9fuVDsdvN1mmFSFFGvFEvM0hksh8oPkuEVSkCalUjh4Xo0GsFVgxOvVQr1+oMoCMxpNpnNdksVm11psYNs4T9gbVTucGFcbncHk9cQS2ukIlxMhgShEOJKSkiUWiOTUAdQ/kA==) the `movsxd` in the codegen and replace it with a `mov` (which could be avoided anyway).

---

In L799 above the condition could be written as 
```diff
- if (startIndex < 0 || startIndex > array.Length)
+ if ((uint)startIndex > array.Length)
```
As any negative number casted to uint will be > int.MaxValue